### ISSUE TITLE
Doc clarification for `MaybeUninit::zeroed()`

### ISSUE THREAD

**Issue by danielocfb** [rust-lang/rust#111608]
### Location

https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.zeroed

> Creates a new MaybeUninit<T> in an uninitialized state, with the memory being filled with 0 bytes. It depends on T whether that already makes for proper initialization. For example, MaybeUninit<usize>::zeroed() is initialized, but MaybeUninit<&'static i32>::zeroed() is not because references must not be null.

### Summary

While perhaps implied by the terminology of "memory being filled with 0 bytes", it would be great to explicitly state whether or not padding bytes are included.

For me it's a bit ambiguous because both `std::mem::zeroed` and `std::mem::MaybeUninit` are parametrized by `T`. Hence, neither really exposes a concept of "raw" bytes to the user, but just the Rust type. Yet if I compare the above documentation to [`std::mem::zeroed`](https://doc.rust-lang.org/std/mem/fn.zeroed.html) it states, among other things:

> This means that, for example, the padding byte in (u8, u16) is not necessarily zeroed.

So there padding bytes are explicitly excluded.

Should we make it more explicit that `std::mem::MaybeUninit::zeroed` also zeroes out padding bytes? (assuming that's the behavior we want; it seems to be the behavior we currently have)

**Comment by quaternic** [comment]
> Should we make it more explicit that `std::mem::MaybeUninit::zeroed` also zeroes out padding bytes? (assuming that's the behavior we want; it seems to be the behavior we currently have)

The difference between
```rust
let zeroed = MaybeUninit::<(u8, u16)>::zeroed();
ptr::copy::<MaybeUninit<(u8, u16)>>(&zeroed as *const _ as *const _, x, 1);
// does not necessarily overwrite the padding byte
```
and
```rust
let zeroed = [0u16; 2];
ptr::copy::<MaybeUninit<(u8, u16)>>(&zeroed as *const _ as *const _, x, 1);
// writes all four bytes
```
seems to imply that `MaybeUninit::<(u8, u16)>::zeroed()` leaves the padding byte undefined. [playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=21ea25c61c80b5fb61973d6a4e3166f2)

**Comment by danielocfb** [comment]
Yeah, I think https://github.com/rust-lang/rust/issues/111616 is looking into that behavior. Sorry if I have misrepresented it, I just peeked at the code and the above was my conclusion.

**Comment by Dylan-DPC** [comment]
This is related to https://github.com/rust-lang/rust/issues/99604
